toolz/dicttoolz.py

```diff
@@ -48,14 +57,20 @@ def merge_with(func, *dicts):
     if len(dicts) == 1 and not isinstance(dicts[0], dict):
         dicts = dicts[0]
 
-    result = {}
+    dict_ = OrderedDict
+    result = OrderedDict()
+    return_ordered_dict = True
     for d in dicts:
+        if return_ordered_dict and not isinstance(d, OrderedDict):
+            result = dict(result)
+            dict_ = dict
+            return_ordered_dict = False
```

Comment:
I wonder if this logic can be pulled out of the loop into a separate loop

```python
if all(isinstance(d, OrderedDict) for d in dicts):
    result = OrderedDict()
else:
    result = dict()
```

This does walk through the dicts twice, but I suspect that this is cheap.  It also requires that we make `dicts` concrete and not lazy, but this was already the case due to how we implement `merge_with`, which brings everything in to memory anyway.  

If this doesn't significantly impact performance then it might be preferred for code simplicity's sake